doc: nRF desktop - Split docs into subpages (#13938)
nrfconnect/sdk-nrf: applications/nrf_desktop/nrf_desktop_integration.rst

.. _nrf_desktop_porting_guide:

Integrating your own hardware
#############################

.. contents::
   :local:
   :depth: 2

> This section describes how to adapt the nRF Desktop application to different hardware, and lists steps required for adding a new board.

Reviewer: Maybe we could link to configuration section from here. I think user must be aware of how nRF Desktop configuration works before starting porting (the sections are no longer next to each other). Something like:
```
.. tip:
   Make sure to get familiar with :ref:`nrf_desktop_configuration` section before porting the application to a new board.

Zephyr support for a board
==========================

Before introducing nRF Desktop application configuration for a given board, you need to ensure that the board is supported in Zephyr.

.. note::
   You can skip this step if your selected board is already supported in Zephyr.

Follow the Zephyr's :ref:`zephyr:board_porting_guide` for detailed instructions related to introducing Zephyr support for a new board.
Make sure that the following conditions are met:

#. Edit the DTS files to make sure they match the hardware configuration.
   Pay attention to the following elements:

   * Pins that are used.
   * Bus configuration for optical sensor.
   * `Changing interrupt priority`_.

#. Edit the board's Kconfig files to make sure they match the required system configuration.
   For example, disable the drivers that will not be used by your device.

.. tip::
   You can define the new board by copying the nRF Desktop reference design files that are the closest match for your hardware and then aligning the configuration to your hardware.
   For example, for gaming mouse use :file:`nrf/boards/arm/nrf52840gmouse_nrf52840`.

> nRF Desktop configuration

Reviewer 2: Question - would it make sense to change it to "Configuring nRF Desktop"/"Configure nRF Desktop" to highlight that these are configuration steps rather than a list of configurations (also to differentiate it from other sections with the same name)?

Reviewer: Consider `nRF Desktop support for a board` (section is next to `Zephyr board support` and describes how to add nRF Desktop specific configuration files for a given board).

Reviewer 2: How about unifying this naming for both sections, while still implying through the naming that there are steps to be performed? E.g., 
1. Adding Zephyr board support 
2. Adding nRF Desktop board support ?

Author: After offline discussion, updated as : nRF Desktop support for a board.

Zephyr recommends to use DTS for hardware configuration (see :ref:`zephyr:dt_vs_kconfig`).
For the new motion sensor configuration to be recognized by DTS, define a dedicated DTS binding.
See :ref:`dt-bindings` for more information, and refer to :file:`dts/bindings/sensor` for binding examples.

.. rst-class:: numbered-step

Configure sensor through DTS
============================

Once binding is defined, it is possible to set the sensor configuration.
This is done by editing the DTS file that describes the board.
For more information, see :ref:`devicetree-intro`.

As an example, take a look at the PMW3360 sensor that already exists in |NCS|.
The following code excerpt is taken from :file:`boards/arm/nrf52840gmouse_nrf52840/nrf52840gmouse_nrf52840.dts`:

.. code-block:: none

   &spi1 {
      compatible = "nordic,nrf-spim";
      status = "okay";
      cs-gpios = <&gpio0 13 0>;

    pinctrl-0 = <&spi1_default_alt>;
    pinctrl-1 = <&spi1_sleep_alt>;
    pinctrl-names = "default", "sleep";
        pmw3360@0 {
          compatible = "pixart,pmw3360";
          reg = <0>;
          irq-gpios = <&gpio0 21 0>;
          spi-max-frequency = <2000000>;
        };
    };

> The communication with PMW3360 is done through the SPI, which makes the sensor a subnode of the SPI bus node.

Author: I will create a separate commit for these fixes. Pekka is yet to review other documents. So, I will wait and apply all the suggestions in a single commit.

Otherwise, the old version is used.

.. _nrf_desktop_bootloader_serial_dfu:

Serial recovery DFU
********************

The serial recovery DFU is a feature of MCUboot and it needs to be enabled in the bootloader's configuration.
For the configuration details, see the :ref:`nrf_desktop_configuring_mcuboot_bootloader` section.

To start the serial recovery DFU, the device should boot into recovery mode, in which the bootloader will be waiting for a new image upload to start.
In the serial recovery DFU mode, the new image is transferred through an USB CDC ACM class instance.
The bootloader overwrites the existing application located on the primary slot with the new application image.
If the transfer is interrupted, the device will not be able to boot the incomplete application, and the image upload must be performed again.

Once the device enters the serial recovery mode, you can use the :ref:`mcumgr <zephyr:device_mgmt>` to:

* Query information about the present image.
* Upload the new image.
  The :ref:`mcumgr <zephyr:device_mgmt>` uses the :file:`zephyr/app_update.bin` update image file.
  It is generated by the build system when building the firmware.

> For example, the following line will start the upload of the new image to the device:

Reviewer 3: ```suggestion
For example, the following line starts the upload of the new image to the device:
```

.. code-block:: console

  mcumgr -t 60 --conntype serial --connstring=/dev/ttyACM0 image upload build-nrf52833dongle_nrf52833/zephyr/app_update.bin

The command assumes that ``/dev/ttyACM0`` serial device is used by the MCUboot bootloader for the serial recovery.

fwupd support
*************

fwupd is an open-source project providing tools and daemon for managing the installation of firmware updates on Linux-based systems.
Together with the `LVFS (Linux Vendor Firmware Service) <LVFS_>`_, it provides a solution for vendors to easily distribute firmware for compatible devices.

> The fwupd tools can communicate with the devices running the nRF Desktop application with the :ref:`nrf_desktop_bootloader_background_dfu` feature enabled.

Reviewer 3: ```suggestion
The fwupd tools can communicate with devices running the nRF Desktop application with the :ref:`nrf_desktop_bootloader_background_dfu` feature enabled.
```

The communication protocols are implemented using plugins.
The plugin associated with the DFU protocol realized through the :ref:`nrf_desktop_config_channel` is branded as ``nordic_hid``.

Adding a new device
===================

The device specifies which protocol is used for the communication by adding the specific device information to the plugin metadata.
For the ``nordic_hid`` plugin, this file is located at `nordic-hid.quirk`_.
The following example shows the information passed to the plugin metadata:

.. code-block:: console

   [HIDRAW\VEN_1915&DEV_52DE]
   Plugin = nordic_hid
   GType = FuNordicHidCfgChannel
   NordicHidBootloader = B0

In this console snippet:

* ``[HIDRAW\VEN_1915&DEV_52DE]`` - This line describes the device instance ID provided by the OS, which identifies the device.
* ``Plugin = nordic_hid`` and ``GType = FuNordicHidCfgChannel`` - These lines set the plugin that the device uses.
* ``NordicHidBootloader`` - This optional line selects the bootloader that the device is running.
  If the device does not have the information about the underlying bootloader, the ``NordicHidBootloader`` option is used to select a proper bootloader type.

>   If there is no information about the bootloader, both in metadata and from the device, the update procedure will fail.

Reviewer 3: ```suggestion
  If there is no information about the bootloader, both in metadata and from the device, the update procedure fails.
```

  The possible values are either ``B0``, ``MCUBOOT`` or ``MCUBOOT+XIP``.

.. note::
   As the ``nordic_hid`` plugin communicates with the device using the Configuration channel, the device update is not allowed through the Serial recovery DFU.

To add a new device, a pull request must be opened to the `fwupd`_ repository with a new entry to the :file:`nordic-hid.quirk` file.

LVFS and update image preparation
=================================

The `LVFS (Linux Vendor Firmware Service) <LVFS_>`_ hosts firmware images that can be downloaded by Linux machines and used by the fwupd tool for the firmware update of compatible devices.
A vendor account is needed to upload a new firmware archive to the site.
Information on how to apply for an account is found at the `LVFS Getting an account`_ website.

The nRF Desktop application DFU image is delivered as a zip package, containing a manifest and one or more binary files used for the update.

> To prepare the image file compatible with the LVFS, the CAB file needs to be prepared.

Reviewer 3: ```suggestion
To prepare an image file compatible with the LVFS, you must prepare a CAB file.
```

The CAB package must contain the DFU package generated by the |NCS|, that is :file:`dfu_application.zip`, plus metadata file with information used by the LVFS.
For more information, see the `LVFS metadata`_ site.

When the CAB archive has been built, it can be uploaded to the LVFS where the CAB archive is verified and signed.
For more information about creating CAB files, signing, and uploading the update package, see the `LVFS Uploading Firmware`_ site.

Upgrading firmware using fwupd
==============================

> Once the update image was uploaded onto the LVFS, the firmware update procedure can be tested on end machines.

Reviewer 3: ```suggestion
Once the update image is uploaded onto the LVFS, you can test the firmware update procedure on end machines.
```
End machines sounds weird. Can we find something better?

Reviewer: Consider:
`Once the update image is uploaded onto the LVFS, you can test the firmware update procedure on HID hosts that support fwupd.`

Complete the following steps:

1. Make sure that the host machine to which the updatable device running the nRF Desktop application is connected has the fwupd tool installed.

> #. Fetch the information about available update images from the LVFS by using the following commands:

Reviewer 3: ```suggestion
#. Fetch the information about available update images from the LVFS using the following commands:
```

   .. code-block:: console

      fwupdmgr refresh
      fwupdmgr get-updates

>    In this console snippet:

Reviewer 3: ```suggestion
```
